text/0000-rpc-queue.md

```diff
@@ -0,0 +1,86 @@
+4. RPC ILPP generated code writes serialized parameters of the RPC via the stream writer
+5. RPC ILPP generated code invokes EndSendXXXXRpc method where the writer used to write serialized parameters is returned back to the RPCQueueContainer
+6. The RPCQueueContainer finalizes the serialized RPC Queue entry
+7. Later in the player loop (game update loop) the RPCQueueContainer processes through all outbound RPCs via RPCQueueProcessing and they are sent individually or batched.
+
+
+
+![](0000-rpc-queue/RCPQueueInboundHigh.png)
+
+1. The NetworkingManager enumerates through all current transport events (in the case of UNET) and adds any received RPC messages to the current inbound QueueHistoryFrame via the RPCQueueContainer (Network PreUpdate Stage).
+2. Later in the PlayerLoop (Network FixedUpdate stage) the RpcQueueContainer enumerates through all inbound RPCs and as long as they are valid will invoke each RPC, NetworkedObject relative, within the associated NetworkedBehaviour component instance initiated by the sender.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+### Outbound RPC Data Flow Pipeline
+![](0000-rpc-queue/OutboundDataFlowPipeline.png)
+
+### Inbound RPC Data Flow Pipeline
+![](0000-rpc-queue/InboundDataFlowPipeline.png)
+
+### RPC Queue Classes
+![](0000-rpc-queue/RPCQueueClasses.png)
+
+- **RPCQueueContainer:** Manages inbound and outbound RPC queues.  Both inbound and outbound queues are byte arrays exposed as streams (i.e. currently BitStreams) and contained within a QueueHistoryFrame.  The number of QueueHistoryFrames is defined by calling the Initialize method and passing the maximum number of history frames.  The total number of frames is MaxFrameHistory + 1, where the additional frame is considered the “current frame” (i.e. it will always maintain the exact number of frames in history).
+- **RPCQueueProcessing:** Currently, this class is instantiated by the RPCQueueContainer during its initialization.  It is highly probable that this class will get absorbed into the RPCQueueContainer.
+- **QueueHistoryFrame:** Container class for handling the management of an RPC queue.  One QueueHistoryFrame instance per data flow pipeline (i.e. inbound and outbound yields two QueueHistoryFrames).
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+This approach does include additional processing overhead and there could be some refactoring depending upon any future transport related work that might occur.  
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+Since there are no current discussions on batching specific packet types together at the transport layer (not recommended) and while Unet provides the ability to switch from immediate to queued mode there were other framework related design factors taken into condiseration (i.e. batched RPCs, invocation at different update stages, etc.) that made this the current best path to take.
```

Comment:
why not comment (at least briefly) what we considered and why this approach suits our needs for now? also, an alternative might be to have `QueueSend()`/`FlushSend()`/etc. APIs on transport itself and expect it to happen under the hood for us, implemented by transport implementors. I'd like to see at least transport-level queuing vs framework-level queuing being mentioned here (especially since we have plans in the future in that direction already, why not note why this particular RFC is more useful than others today but potentially not tomorrow?).

Reply:
Added a blurb on this in the most recent refactor.  I held back on going into any specific details as we still need some additional clarity regarding what we are going to expect from the default transport used with MLAPI (i.e. what the default Unity transport will end up being) and any changes to the framework transport interface that may be needed.

Reply:
The most recent version should address this area further.